gensim/models/word2vec.py

```diff
@@ -582,233 +295,1357 @@ def __init__(self, sentences=None, corpus_file=None, size=100, alpha=0.025, wind
-        job_queue = Queue(maxsize=queue_factor * self.workers)
-        progress_queue = Queue(maxsize=(queue_factor + 1) * self.workers)
-
-        workers = [threading.Thread(target=worker_loop) for _ in range(self.workers)]
-        for thread in workers:
-            thread.daemon = True  # make interrupting the process with ctrl+c easier
-            thread.start()
+        # Since no sentences are provided, this is to control the corpus_count.
+        self.corpus_count = corpus_count or 0
+        self.raw_vocab = raw_vocab
 
-        sentence_count = 0
-        sentence_scores = matutils.zeros_aligned(total_sentences, dtype=REAL)
+        # trim by min_count & precalculate downsampling
+        report_values = self.prepare_vocab(keep_raw_vocab=keep_raw_vocab, trim_rule=trim_rule, update=update)
+        report_values['memory'] = self.estimate_memory(vocab_size=report_values['num_retained_words'])
+        self.prepare_weights(update=update)  # build tables & arrays
 
-        push_done = False
-        done_jobs = 0
+    def _scan_vocab(self, sentences, progress_per, trim_rule):
+        sentence_no = -1
+        total_words = 0
+        min_reduce = 1
+        vocab = defaultdict(int)
+        checked_string_types = 0
+        for sentence_no, sentence in enumerate(sentences):
+            if not checked_string_types:
+                if isinstance(sentence, string_types):
+                    logger.warning(
+                        "Each 'sentences' item should be a list of words (usually unicode strings). "
+                        "First item here is instead plain %s.",
+                        type(sentence)
+                    )
+                checked_string_types += 1
+            if sentence_no % progress_per == 0:
+                logger.info(
+                    "PROGRESS: at sentence #%i, processed %i words, keeping %i word types",
+                    sentence_no, total_words, len(vocab)
+                )
+            for word in sentence:
+                vocab[word] += 1
+            total_words += len(sentence)
+
+            if self.max_vocab_size and len(vocab) > self.max_vocab_size:
+                utils.prune_vocab(vocab, min_reduce, trim_rule=trim_rule)
+                min_reduce += 1
+
+        corpus_count = sentence_no + 1
+        self.raw_vocab = vocab
+        return total_words, corpus_count
+
+    def scan_vocab(self, corpus_iterable=None, corpus_file=None, progress_per=10000, workers=None, trim_rule=None):
+        logger.info("collecting all words and their counts")
+        if corpus_file:
+            corpus_iterable = LineSentence(corpus_file)
+
+        total_words, corpus_count = self._scan_vocab(corpus_iterable, progress_per, trim_rule)
+
+        logger.info(
+            "collected %i word types from a corpus of %i raw words and %i sentences",
+            len(self.raw_vocab), total_words, corpus_count
+        )
+
+        return total_words, corpus_count
+
+    def prepare_vocab(
```

Comment:
This method is quite long. Is there a good way to break it up?

Reply:
Currently, it's a multi-author mess I've only moved, not improved in this PR. In near future, I want to refactor the `.build_vocab()` steps, which should slim/split this & related methods.